src/backend/gl/src/command.rs

```diff
@@ -474,54 +474,57 @@ impl RawCommandBuffer {
-                                    Command::ClearBufferColorU(0, unsafe { cv.color.uint32 })
-                                }
-                                ChannelType::Sint => {
-                                    Command::ClearBufferColorI(0, unsafe { cv.color.int32 })
-                                }
-                            };
-
-                            return Some(cmd);
-                        }
+                        assert!(clear.index >= glow::COLOR_ATTACHMENT0 && clear.index <= glow::COLOR_ATTACHMENT31);
+                        assert_eq!(attachment.ops.load, pass::AttachmentLoadOp::Clear);
+
+                        let channel = view_format.base_format().1;
+
+                        let cmd = match channel {
+                            ChannelType::Unorm
+                            | ChannelType::Snorm
+                            | ChannelType::Ufloat
+                            | ChannelType::Sfloat
+                            | ChannelType::Srgb
+                            | ChannelType::Uscaled
+                            | ChannelType::Sscaled => {
+                                Command::ClearBufferColorF(clear.index - glow::COLOR_ATTACHMENT0, unsafe { clear.value.color.float32 })
+                            }
+                            ChannelType::Uint => {
+                                Command::ClearBufferColorU(clear.index - glow::COLOR_ATTACHMENT0, unsafe { clear.value.color.uint32 })
+                            }
+                            ChannelType::Sint => {
+                                Command::ClearBufferColorI(clear.index - glow::COLOR_ATTACHMENT0, unsafe { clear.value.color.int32 })
+                            }
+                        };
+
+                        return Some(cmd);
                     }
 
                     // Clear depth-stencil target
-                    let depth = if view_format.is_depth() {
-                        clear.value.map(|cv| unsafe { cv.depth_stencil.depth })
+                    let depth = if view_format.is_depth() && attachment.ops.load == pass::AttachmentLoadOp::Clear {
+                        Some(unsafe { clear.value.depth_stencil.depth })
                     } else {
                         None
                     };
 
-                    let stencil = if view_format.is_stencil() {
-                        clear.stencil_value
+                    let stencil = if view_format.is_stencil() && attachment.stencil_ops.load == pass::AttachmentLoadOp::Clear {
+                        Some(unsafe { clear.value.depth_stencil.stencil })
                     } else {
                         None
                     };
 
                     if depth.is_some() || stencil.is_some() {
                         return Some(Command::ClearBufferDepthStencil(depth, stencil));
                     }
+                }
+                None
+            })
+            .collect::<Vec<_>>();
```

Comment:
do we have to collect it here?

Reply:
We need access by index

Reply:
where are we accessing `clear_cmds` by index?

Reply:
We don't, I confused it with another piece of code

Reply:
Apparently we need to let go of `self` borrow, so collecting is unavoidable.

Reply:
The `self` is only borrowed for `self.cur_subpass`? In this case, we can copy it into a variable and not hold `self`

Reply:
No. `self` is borrowed for `state` which owns collections we iterate over.